### ISSUE TITLE
Issue with compiling LIB

### ISSUE THREAD

**Issue by ricardoalcantara** [GodotNativeTools/godot-cpp#233]
I'm trying to compile a lib which is a simple interface to mongocxx driver that access mongodb, I could do a simple select and print inside Godot, them I made those classes separated in another project and worked it till it returns values, now I am trying to put it inside the GDNative module cannot compile it, and since it lacks few documentation I am creating a issue here, sorry if I miss something.

It's open source, this is exactly the code I am trying to compile:
https://github.com/ricardoalcantara/godot-mongodb-gdnative

This is the error I am having

```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
g++ -o src/gd_mongodb.os -c -fPIC -g3 -Og -std=c++17 -pthread -fPIC -DMONGOCXX_STATIC -DBSONCXX_STATIC -DMONGOC_STATIC -DBSON_STATIC -I. -Igodot-cpp/godot_headers -Igodot-cpp/include -Igodot-cpp/include/core -Igodot-cpp/include/gen -I/usr/local/include/mongocxx/v_noabi -I/usr/local/include/bsoncxx/v_noabi -I/usr/local/include/libmongoc-1.0 -I/usr/local/include/libbson-1.0 -Isrc src/gd_mongodb.cpp
g++ -o src/gd_mongodb_collection.os -c -fPIC -g3 -Og -std=c++17 -pthread -fPIC -DMONGOCXX_STATIC -DBSONCXX_STATIC -DMONGOC_STATIC -DBSON_STATIC -I. -Igodot-cpp/godot_headers -Igodot-cpp/include -Igodot-cpp/include/core -Igodot-cpp/include/gen -I/usr/local/include/mongocxx/v_noabi -I/usr/local/include/bsoncxx/v_noabi -I/usr/local/include/libmongoc-1.0 -I/usr/local/include/libbson-1.0 -Isrc src/gd_mongodb_collection.cpp
g++ -o src/gd_mongodb_database.os -c -fPIC -g3 -Og -std=c++17 -pthread -fPIC -DMONGOCXX_STATIC -DBSONCXX_STATIC -DMONGOC_STATIC -DBSON_STATIC -I. -Igodot-cpp/godot_headers -Igodot-cpp/include -Igodot-cpp/include/core -Igodot-cpp/include/gen -I/usr/local/include/mongocxx/v_noabi -I/usr/local/include/bsoncxx/v_noabi -I/usr/local/include/libmongoc-1.0 -I/usr/local/include/libbson-1.0 -Isrc src/gd_mongodb_database.cpp
In file included from src/gd_mongodb_collection.h:4:0,
                 from src/gd_mongodb_collection.cpp:3:
godot-cpp/include/core/Godot.hpp: In instantiation of 'const char* godot::___get_method_class_name(R (T::*)(args ...)) [with T = godot::GDMongoDBCollection; R = void; args = {godot::Dictionary}]':
godot-cpp/include/core/Godot.hpp:246:130:   required from 'void godot::register_method(const char*, M, godot_method_rpc_mode) [with M = void (godot::GDMongoDBCollection::*)(godot::Dictionary)]'
src/gd_mongodb_collection.cpp:8:63:   required from here
godot-cpp/include/core/Godot.hpp:149:28: error: '___get_type_name' is not a member of 'godot::GDMongoDBCollection'
  return T::___get_type_name();
         ~~~~~~~~~~~~~~~~~~~^~
godot-cpp/include/core/Godot.hpp: In instantiation of 'const char* godot::___get_method_class_name(R (T::*)(args ...)) [with T = godot::GDMongoDBCollection; R = godot::Dictionary; args = {godot::Dictionary}]':
godot-cpp/include/core/Godot.hpp:246:130:   required from 'void godot::register_method(const char*, M, godot_method_rpc_mode) [with M = godot::Dictionary (godot::GDMongoDBCollection::*)(godot::Dictionary)]'
src/gd_mongodb_collection.cpp:9:59:   required from here
godot-cpp/include/core/Godot.hpp:149:28: error: '___get_type_name' is not a member of 'godot::GDMongoDBCollection'
scons: *** [src/gd_mongodb_collection.os] Error 1
In file included from src/gd_mongodb.h:4:0,
                 from src/gd_mongodb.cpp:5:
godot-cpp/include/core/Godot.hpp: In instantiation of 'void godot::_WrappedMethod<T, R, As>::apply(godot::Variant*, T*, godot::Variant**, godot::__Sequence<I ...>) [with int ...I = {0}; T = godot::GDMongoDB; R = godot::GDMongoDBConnection; As = {godot::String}]':
godot-cpp/include/core/Godot.hpp:208:2:   required from 'godot_variant godot::__wrapped_method(godot_object*, void*, void*, int, godot_variant**) [with T = godot::GDMongoDB; R = godot::GDMongoDBConnection; As = {godot::String}; godot_object = void]'
godot-cpp/include/core/Godot.hpp:223:9:   required from 'godot_variant (* godot::___get_wrapper_function(R (T::*)(As ...)))(godot_object*, void*, void*, int, godot_variant**) [with T = godot::GDMongoDB; R = godot::GDMongoDBConnection; As = {godot::String}; godot::__godot_wrapper_method = godot_variant (*)(void*, void*, void*, int, godot_variant**); godot_object = void]'
godot-cpp/include/core/Godot.hpp:241:65:   required from 'void godot::register_method(const char*, M, godot_method_rpc_mode) [with M = godot::GDMongoDBConnection (godot::GDMongoDB::*)(godot::String)]'
src/gd_mongodb.cpp:10:61:   required from here
godot-cpp/include/core/Godot.hpp:183:8: error: no match for 'operator=' (operand types are 'godot::Variant' and 'godot::GDMongoDBConnection')
   *ret = (obj->*f)(_ArgCast<As>::_arg_cast(*args[I])...);
   ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from godot-cpp/include/core/Dictionary.hpp:4:0,
                 from godot-cpp/include/core/CoreTypes.hpp:10,
                 from godot-cpp/include/core/Godot.hpp:11,
                 from src/gd_mongodb.h:4,
                 from src/gd_mongodb.cpp:5:
godot-cpp/include/core/Variant.hpp:196:11: note: candidate: godot::Variant& godot::Variant::operator=(const godot::Variant&)
  Variant &operator=(const Variant &v);
           ^~~~~~~~
godot-cpp/include/core/Variant.hpp:196:11: note:   no known conversion for argument 1 from 'godot::GDMongoDBConnection' to 'const godot::Variant&'
scons: *** [src/gd_mongodb.os] Error 1
In file included from src/gd_mongodb.h:4:0,
                 from src/gd_mongodb_database.cpp:1:
godot-cpp/include/core/Godot.hpp: In instantiation of 'void godot::_WrappedMethod<T, R, As>::apply(godot::Variant*, T*, godot::Variant**, godot::__Sequence<I ...>) [with int ...I = {0}; T = godot::GDMongoDBDatabase; R = godot::GDMongoDBCollection; As = {godot::String}]':
godot-cpp/include/core/Godot.hpp:208:2:   required from 'godot_variant godot::__wrapped_method(godot_object*, void*, void*, int, godot_variant**) [with T = godot::GDMongoDBDatabase; R = godot::GDMongoDBCollection; As = {godot::String}; godot_object = void]'
godot-cpp/include/core/Godot.hpp:223:9:   required from 'godot_variant (* godot::___get_wrapper_function(R (T::*)(As ...)))(godot_object*, void*, void*, int, godot_variant**) [with T = godot::GDMongoDBDatabase; R = godot::GDMongoDBCollection; As = {godot::String}; godot::__godot_wrapper_method = godot_variant (*)(void*, void*, void*, int, godot_variant**); godot_object = void]'
godot-cpp/include/core/Godot.hpp:241:65:   required from 'void godot::register_method(const char*, M, godot_method_rpc_mode) [with M = godot::GDMongoDBCollection (godot::GDMongoDBDatabase::*)(godot::String)]'
src/gd_mongodb_database.cpp:6:69:   required from here
godot-cpp/include/core/Godot.hpp:183:8: error: no match for 'operator=' (operand types are 'godot::Variant' and 'godot::GDMongoDBCollection')
   *ret = (obj->*f)(_ArgCast<As>::_arg_cast(*args[I])...);
   ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from godot-cpp/include/core/Dictionary.hpp:4:0,
                 from godot-cpp/include/core/CoreTypes.hpp:10,
                 from godot-cpp/include/core/Godot.hpp:11,
                 from src/gd_mongodb.h:4,
                 from src/gd_mongodb_database.cpp:1:
godot-cpp/include/core/Variant.hpp:196:11: note: candidate: godot::Variant& godot::Variant::operator=(const godot::Variant&)
  Variant &operator=(const Variant &v);
           ^~~~~~~~
godot-cpp/include/core/Variant.hpp:196:11: note:   no known conversion for argument 1 from 'godot::GDMongoDBCollection' to 'const godot::Variant&'
scons: *** [src/gd_mongodb_database.os] Error 1
scons: building terminated because of errors.

```

**Comment by bruvzg** [comment]
1. >  error: '___get_type_name' is not a member of 'godot::GDMongoDBCollection'

`GODOT_CLASS(GDMongoDBCollection, Reference);` line is missing from `src/gd_mongodb_collection.h`

2. You can't use constructors in gdnative all init should be done in `_init` method.

3. > error: no match for 'operator=' (operand types are 'godot::Variant' and 'godot::GDMongoDBConnection')

You should return classes by reference (note there is a bug in current library version and references are leaking #215).

```c++
//in gd_mongodb.cpp
Ref<GDMongoDBConnection> GDMongoDB::GetConnection(String uri) {
    Ref<GDMongoDBConnection> ret = Ref<GDMongoDBConnection>::__internal_constructor(GDMongoDBConnection::_new()); //hack to prevent leak

    //Ref<GDMongoDBConnection> ret = Ref<GDMongoDBConnection>(GDMongoDBConnection::_new()); <-- should work but leaks

    //Ref<GDMongoDBConnection> ret;
    //ret.instance(); <-- should work but leaks

    ret->OpenURI(uri);
    return ret;
}

//in gd_mongodb_connection.cpp
GDMongoDBConnection::GDMongoDBConnection() { /*UNUSED*/}

void GDMongoDBConnection::GDMongoDBConnection::OpenURI(String uri) {
	mongocxx::uri _uri{uri.alloc_c_string()};
	_client = new mongocxx::client{_uri};
}
```

instead of

```c++
//in gd_mongodb.cpp
GDMongoDBConnection GDMongoDB::GetConnection(String uri) {
    return GDMongoDBConnection(uri);
}

//in gd_mongodb_connection.cpp
GDMongoDBConnection::GDMongoDBConnection(String uri) {
	mongocxx::uri _uri{uri.alloc_c_string()};
	_client = new mongocxx::client{_uri};
}
```

4. Register all classes in `gd_library.cpp`

**Comment by ricardoalcantara** [comment]
First of all, Thank you very much @bruvzg ! And sorry that I am not very skilled on C++, I have always worked at kinder-garden C# and other sweet heart languages. And after a while seen a lot of errors I don't know how to read properly I felt lost.

I have made those changes you told me and one or two more I caught and it compiled at first!

I will keep a track on that issue till it got fixed and I will update my code!

Thanks again.
